opentimelineio/algorithms/filter.py

```diff
@@ -0,0 +1,285 @@
+
+    return mutable_object
+
+
+def _safe_parent(child):
+    if hasattr(child, 'parent'):
+        return child.parent()
+    return None
+
+
+def filtered_with_sequence_context(
+    input_object,
+    reduce_fn,
+    types_to_prune=None,
+    types_to_filter=None,
+):
+    """Filter input_object with reduce_fn to make a new copy of object.
+
+    reduce_fn will be called with the objects from input_object, in the form
+    reduce_fn(prev_item, child, next_item) where prev_item and next_item are
+    the result of calling child.parent().neighbors_of(child) *from the original
+    sequence*, or None if the item is not in a sequence.
+
+    If reduce_fn returns an object, that object will be inserted into the
+    hierarchy, replacing whichever object was passed in to it.  If it returns
+    None, that object will be pruned, and all of its children will be omitted
+    from iteration.  If the function returns a tuple, that tuple will be
+    inserted into the hierarchy.  If an object is pruned, it will still be
+    passed into the next iteration as the prev_item call for the next item.
+
+    If you have a track: [A,B,C]
+
+    If your reduce_fn is:
+        def fn(prev_item, thing, next_item):
+            if prev_item.name == A:
+                return D # new clip
+            else:
+                return thing
+
+    if filtered_with_sequence_context(track, fn) is called, the order will be:
+        fn(None, A, B) => A
+        fn(A, B, C) => D
+        fn(B, C, D) => C # !! note that it was passed B instead of D.
+    So the result is:
+        [A, D, C]
+
+    If your reduce_fn is:
+        def fn(prev_item, thing, next_item):
+            if prev_item.name == A:
+                return None # prune the clip
+            else:
+                return thing
+
+    if filtered_with_sequence_context(track, fn) is called, the order will be:
+        fn(None, A, B) => A
+        fn(A, B, C) => None
+        fn(B, C, D) => C # !! note that it was passed B instead of D.
+    So the result is:
+        [A, C]
+
+    If your reduce_fn is:
+        def fn(prev_item, thing, next_item):
+            if prev_item.name == A:
+                return (D, E) # tuple of new clips
+            else:
+                return thing
+
+    if filtered_with_sequence_context(track, fn) is called, the order will be:
+        fn(None, A, B) => A
+        fn(A, B, C) => (D, E)
+        fn(B, C, D) => C # !! note that it was passed B instead of D.
+    So the result is:
+        [A, D, E, C]
+
+    Additionally, types_to_prune and types_to_filter let you skip or
+    prune classes without having to build logic into your function.
+
+    The order of operation is:
+        for each thing to iterate on:
+            1 if it isinstance of anything in the prune list, prune it.
+            2 if it is not isinstance of anything in the types_to_filter,
+                or the types_to_filter is None, pass it through, without
+                giving it to your function
```

Comment:
My read of this is that if `types_to_filter` is `None` (which is the default), then we'll pass it through without giving it to the function - which I don't think is the case...?  Instead what about replacing `or the types_to_filter is None` with `and the types_to_filter is not None`?